pkg/storage/utils/indexer/indexer.go

```diff
@@ -0,0 +1,432 @@
+func (i *Indexer) FindByPartial(t interface{}, field string, pattern string) ([]string, error) {
+	typeName := getTypeFQN(t)
+
+	i.mu.RLock(typeName)
+	defer i.mu.RUnlock(typeName)
+
+	resultPaths := make([]string, 0)
+	if fields, ok := i.indices[typeName]; ok {
+		for _, idx := range fields.IndicesByField[strcase.ToCamel(field)] {
+			res, err := idx.Search(pattern)
+			if err != nil {
+				if errors.IsNotFoundErr(err) {
+					continue
+				}
+
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			resultPaths = append(resultPaths, res...)
+
+		}
+	}
+
+	result := make([]string, 0, len(resultPaths))
+	for _, v := range resultPaths {
+		result = append(result, path.Base(v))
+	}
+
+	return result, nil
+
+}
+
+// Update updates all indexes on a value <from> to a value <to>.
+func (i *Indexer) Update(from, to interface{}) error {
+	typeNameFrom := getTypeFQN(from)
+
+	i.mu.Lock(typeNameFrom)
+	defer i.mu.Unlock(typeNameFrom)
+
+	if typeNameTo := getTypeFQN(to); typeNameFrom != typeNameTo {
+		return fmt.Errorf("update types do not match: from %v to %v", typeNameFrom, typeNameTo)
+	}
+
+	if fields, ok := i.indices[typeNameFrom]; ok {
+		for fName, indices := range fields.IndicesByField {
+			oldV, err := valueOf(from, option.IndexByField(fName))
+			if err != nil {
+				return err
+			}
+			newV, err := valueOf(to, option.IndexByField(fName))
+			if err != nil {
+				return err
+			}
+			pkVal, err := valueOf(from, option.IndexByField(fields.PKFieldName))
+			if err != nil {
+				return err
+			}
+			for _, idx := range indices {
+				if oldV == newV {
+					continue
+				}
+				if oldV == "" {
+					if _, err := idx.Add(pkVal, newV); err != nil {
+						return err
+					}
+					continue
+				}
+				if newV == "" {
+					if err := idx.Remove(pkVal, oldV); err != nil {
+						return err
+					}
+					continue
+				}
+				if err := idx.Update(pkVal, oldV, newV); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// Query parses an OData query into something our indexer.Index understands and resolves it.
+func (i *Indexer) Query(ctx context.Context, t interface{}, q string) ([]string, error) {
+	query, err := godata.ParseFilterString(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	tree := newQueryTree()
+	if err := buildTreeFromOdataQuery(query.Tree, &tree); err != nil {
+		return nil, err
+	}
+
+	results := make([]string, 0)
+	if err := i.resolveTree(t, &tree, &results); err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+// t is used to infer the indexed field names. When building an index search query, field names have to respect Golang
+// conventions and be in PascalCase. For a better overview on this contemplate reading the reflection package under the
+// indexer directory. Traversal of the tree happens in a pre-order fashion.
+// TODO implement logic for `and` operators.
+func (i *Indexer) resolveTree(t interface{}, tree *queryTree, partials *[]string) error {
```

Comment:
I had to try it because I wasn't sure myself, but when the slice is given by value you can alter the elements being pointed at by both the original slice and the copy, but you can't alter the original slice itself, i.e. adding/removing elements is not possible. See https://go.dev/play/p/Mk4IkD0dKz9